graphics/blit/blit-alpha.h

```diff
@@ -0,0 +1,221 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "graphics/blit.h"
+
+namespace Graphics {
+
+class BlendBlitImpl_Base {
+	friend class BlendBlit;
+protected:
+
+struct AlphaBlend {
+	static inline void normal(const byte *in, byte *out, const byte ca, const byte cr, const byte cg, const byte cb) {
+		uint32 ina = in[BlendBlit::kAIndex] * ca >> 8;
+
+		if (ina != 0) {
+			uint outb = (out[BlendBlit::kBIndex] * (255 - ina) >> 8);
+			uint outg = (out[BlendBlit::kGIndex] * (255 - ina) >> 8);
+			uint outr = (out[BlendBlit::kRIndex] * (255 - ina) >> 8);
+
+			out[BlendBlit::kAIndex] = 255;
+			out[BlendBlit::kBIndex] = outb + (in[BlendBlit::kBIndex] * ina * cb >> 16);
+			out[BlendBlit::kGIndex] = outg + (in[BlendBlit::kGIndex] * ina * cg >> 16);
+			out[BlendBlit::kRIndex] = outr + (in[BlendBlit::kRIndex] * ina * cr >> 16);
+		}
+	}
+};
+
+struct MultiplyBlend {
+	static inline void normal(const byte *in, byte *out, const byte ca, const byte cr, const byte cg, const byte cb) {
+		uint32 ina = in[BlendBlit::kAIndex] * ca >> 8;
+
+		if (ina != 0) {
+			out[BlendBlit::kBIndex] = out[BlendBlit::kBIndex] * ((in[BlendBlit::kBIndex] * cb * ina) >> 16) >> 8;
+			out[BlendBlit::kGIndex] = out[BlendBlit::kGIndex] * ((in[BlendBlit::kGIndex] * cg * ina) >> 16) >> 8;
+			out[BlendBlit::kRIndex] = out[BlendBlit::kRIndex] * ((in[BlendBlit::kRIndex] * cr * ina) >> 16) >> 8;
+		}
+	}
+};
+
+struct OpaqueBlend {
+	static inline void normal(const byte *in, byte *out, const byte ca, const byte cr, const byte cg, const byte cb) {
+		*(uint32 *)out = *(const uint32 *)in | BlendBlit::kAModMask;
+	}
+};
+
+struct BinaryBlend {
+	static inline void normal(const byte *in, byte *out, const byte ca, const byte cr, const byte cg, const byte cb) {
+		uint32 pix = *(const uint32 *)in;
+		int a = in[BlendBlit::kAIndex];
+
+		if (a != 0) {   // Full opacity (Any value not exactly 0 is Opaque here)
+			*(uint32 *)out = pix;
+			out[BlendBlit::kAIndex] = 0xFF;
+		}
+	}
+};
+
+struct AdditiveBlend {
+	static inline void normal(const byte *in, byte *out, const byte ca, const byte cr, const byte cg, const byte cb) {
+		uint32 ina = in[BlendBlit::kAIndex] * ca >> 8;
+
+		if (ina != 0) {
+			out[BlendBlit::kBIndex] = out[BlendBlit::kBIndex] + ((in[BlendBlit::kBIndex] * cb * ina) >> 16);
+			out[BlendBlit::kGIndex] = out[BlendBlit::kGIndex] + ((in[BlendBlit::kGIndex] * cg * ina) >> 16);
+			out[BlendBlit::kRIndex] = out[BlendBlit::kRIndex] + ((in[BlendBlit::kRIndex] * cr * ina) >> 16);
+		}
+	}
+};
+
+struct SubtractiveBlend {
+	static inline void normal(const byte *in, byte *out, const byte ca, const byte cr, const byte cg, const byte cb) {
```

Comment:
As I understand it, those are used for trailing pixels if the width in bytes isn't cleanly divisible by the SIMD type, so it's still needed for the optimized version.

It should probably also be used to avoid penalties relating to unaligned memory access, but that's something to think about later.

Reply:
You're absolutely correct, I completely missed that usage 🤦